File: crates/sy-hwprobe/src/tpu.rs
//! Google TPU detection via `/sys/class/accel` sysfs.

use crate::types::AcceleratorDevice;
use std::fs;
use std::path::Path;

/// HBM per chip in GB by TPU version.
fn hbm_gb(version: &str) -> u64 {
    match version {
        "v4" => 32,
        "v5e" => 16,
        "v5p" => 95,
        _ => 16,
    }
}

pub fn probe() -> Vec<AcceleratorDevice> {
    let accel_dir = Path::new("/sys/class/accel");
    if !accel_dir.exists() {
        return Vec::new();
    }

    let entries = match fs::read_dir(accel_dir) {
        Ok(e) => e,
        Err(_) => return Vec::new(),
    };

    let mut devices = Vec::new();

    for entry in entries.flatten() {
        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        if !name_str.starts_with("accel") {
            continue;
        }

        let device_path = accel_dir.join(&*name_str).join("device");

        // Skip AMD XDNA devices (handled in npu module)
        if let Ok(target) = fs::read_link(device_path.join("driver")) {
            if target.to_string_lossy().contains("amdxdna") {
                continue;
            }
        }

        // Read TPU version
        let version = fs::read_to_string(device_path.join("tpu_version"))
            .ok()
            .map(|s| {
                let s = s.trim().to_lowercase();
                if s.contains("v5p") {
                    "v5p"
                } else if s.contains("v5e") || s.contains("v5litepod") {
                    "v5e"
                } else if s.contains("v4") {
                    "v4"
                } else {
                    "v5e"
                }
            })
            .unwrap_or("v5e")
            .to_string();

        // Read chip count
        let chip_count: u64 = fs::read_to_string(device_path.join("chip_count"))
            .ok()
            .and_then(|s| s.trim().parse().ok())
            .unwrap_or(1);

        let total_mb = hbm_gb(&version) * chip_count * 1024;
        let idx = devices.len() as u32;

        let mut dev = AcceleratorDevice::new(
            &format!("Google TPU {version} ({chip_count} chips)"),
            "google",
            "tpu",
        );
        dev.index = idx;
        dev.vram_total_mb = total_mb;
        dev.vram_free_mb = total_mb;
        dev.driver_version = "libtpu".into();
        dev.compute_capability = Some(format!("TPU {version}"));
        dev.tpu_available = true;
        devices.push(dev);
    }

    devices
}

File: packages/core/src/ai/accelerator/tpu.ts
/**
 * TPU family probes — Google TPU v4/v5e/v5p.
 *
 * Detection via /dev/accel* + /sys/class/accel sysfs, matching ai-hwaccel
 * and Ifran detection approaches.
 */

import type { AcceleratorDevice } from './types.js';

/** HBM per chip in GB by TPU version (matching ai-hwaccel TpuVersion::hbm_per_chip_bytes). */
const TPU_HBM_GB: Record<string, number> = {
  v4: 32,
  v5e: 16,
  v5p: 95,
};

/** Probe Google TPU via /sys/class/accel sysfs. */
export async function probeTpu(): Promise<AcceleratorDevice[]> {
  try {
    const { readdir, readFile, readlink } = await import('node:fs/promises');
    const { existsSync } = await import('node:fs');
    const { join } = await import('node:path');

    if (!existsSync('/sys/class/accel')) return [];

    let accelEntries: string[];
    try {
      accelEntries = await readdir('/sys/class/accel');
    } catch {
      return [];
    }

    const devices: AcceleratorDevice[] = [];

    for (const entry of accelEntries) {
      if (!entry.startsWith('accel')) continue;
      const devicePath = join('/sys/class/accel', entry, 'device');

      // Skip AMD XDNA devices (handled in npu.ts)
      try {
        const driverTarget = await readlink(join(devicePath, 'driver'));
        if (driverTarget.includes('amdxdna')) continue;
      } catch {
        // No driver link — may still be TPU
      }

      // Read TPU version from sysfs
      let version = 'v5e';
      try {
        const verStr = (await readFile(join(devicePath, 'tpu_version'), 'utf-8')).trim();
        if (verStr.includes('v5p')) version = 'v5p';
        else if (verStr.includes('v5e') || verStr.includes('v5litepod')) version = 'v5e';
        else if (verStr.includes('v4')) version = 'v4';
      } catch {
        // Default to v5e
      }

      // Read chip count
      let chipCount = 1;
      try {
        const countStr = (await readFile(join(devicePath, 'chip_count'), 'utf-8')).trim();
        chipCount = parseInt(countStr, 10) || 1;
      } catch {
        // Default to 1
      }

      const totalMb = (TPU_HBM_GB[version] ?? 16) * chipCount * 1024;

      devices.push({
        index: devices.length,
        name: `Google TPU ${version} (${chipCount} chips)`,
        vendor: 'google',
        family: 'tpu',
        vramTotalMb: totalMb,
        vramUsedMb: 0,
        vramFreeMb: totalMb,
        utilizationPercent: 0,
        temperatureCelsius: null,
        driverVersion: 'libtpu',
        computeCapability: `TPU ${version}`,
        cudaAvailable: false,
        rocmAvailable: false,
        tpuAvailable: true,
      });
    }

    return devices;
  } catch {
    return [];
  }
}
